### ISSUE TITLE
Make some status attributes back private

### ISSUE THREAD

**Issue by rizar** [bartvm/blocks#565]
After the recent log refactoring my status is printed that long:

```
	 batch_interrupt_received: False
	 epoch_interrupt_received: True
	 epoch_started: True
	 epochs_done: 0
	 iterations_done: 22
	 received_first_batch: True
	 training_started: True
```

I think that everything except for `iterations_done` and `epochs_done` deserves an underscore in order to be not printed. I skipped that when reviewing the PR unfortunately, but this look quite ugly.

**Comment by bartvm** [comment]
I'd prefer a solution where the Printing extension itself decides what to
print or not. What's "private" or not can very much depend on the extension.

**Comment by rizar** [comment]
Do you want to have a bunch of ifs in `Printing`? That would be ugly, I think.

There should be some basic annotation of significance of log entries akin to DEBUG, INFO, etc.. If we do not want to fiddle with separate annotations, let's keep using our convention that an underscore is added to those records that by the opinion of the code author would be of low interest for the user.

**Comment by bartvm** [comment]
You don't need ifs, just a class attribute with a list of keys that you
don't want printed by default, and maybe a keyword argument to turn this
filter off. I think we should keep the log simple, but we also shouldn't
move printing logic into the log, so I think a filter list is a good option.

**Comment by rizar** [comment]
But I consider this ugly to put this information into `Printing`. `Printing` and any other extension of its kind should not know names of all possible records of low significance, there should be a common way for them to identify such records, and name mangling is the cheapest option possible.

> I think we should keep the log simple, but we also shouldn't move printing logic into the log, so I think a filter list is a good option.

I do not suggest to make any changes in the log though, neither do I propose to make log aware of printing details. What I am saying is: there should be a convention for log record names that makes it possible for an extension (or main loop) that writes to the log (or status) to declare that a particular record is mainly for internal/debugging usage. That is I propose to change the names back from `received_first_batch` to `_received_first_batch` and etc.

**Comment by bartvm** [comment]
I just don't think that they are "records of low significance", they are "records that shouldn't be printed", because for other extensions `epoch_started` could be the most significant one. That logic is specific to the `Printing` extension then. To make the parallel with the `logging` module: It's the handlers that decide what to do with messages, and in this case the handlers should maybe just be integrated into the extensions.

Basically, I just think that the distinction of "internal/debugging" records and other records is a bit arbitrary. Right now, the distinction is really "what should be printed and what shouldn't", but once we start having more extensions, that distinction might become very blurry, so I'd rather not make it at all.

**Comment by rizar** [comment]
> because for other extensions `epoch_started` could be the most significant one.

I can not think of any real example. And I very much doubt that there is a point for any extension to access `received_first_batch` and even `batch_interrupt_received`. 

To make the parallel with the logging module, every `logger.smth` call declares a logging level for the message which is afterwards interpreted by handlers. The only logging level different from the default one that we so far have is "starts with underscore", and I believe that many of the status attributes such as `received_first_batch` should be logged at this level.

> Basically, I just think that the distinction of "internal/debugging" records and other records is a bit arbitrary. Right now, the distinction is really "what should be printed and what shouldn't", but once we start having more extensions, that distinction might become very blurry, so I'd rather not make it at all.

I do not find distinction between `received_first_batch` and `iterations_done` blurry.

**Comment by bartvm** [comment]
Any extensions that wants to show whether an experiment was ended by the user or not will need to access `epoch_interrupt_received` and `batch_interrupt_received` and `epoch_started` can be useful for a plotting extension that wants to draw lines at the points in time where a new epoch begun.

**Comment by rizar** [comment]
> `epoch_started` can be useful for a plotting extension that wants to draw lines at the points in time where a new epoch begun. 

There are `before_epoch` and `after_epoch` for that. 

I agree that a very rare exception can actually be interested in the reason why training should be finished. But usage of a record by other extensions is not where I propose to draw the line. By analogy with logging I think that there is certain measure of how much a record is interesting for user. This measure is agnostic to the way logging information is presented to the user: printing as in `Printing`, pseudo-graphical interface, web-browser, etc.. I do not want to make each and every extension that renders the log content to have a hard-coded default list of attributes that should be ignored. As we support no meta-information for log and status records (and arguably should not care about that so far) the only way a publisher can declare that a certain log record is a bad candidate for rendering is name mangling.

I think that regardless of the frontend used to display training progress people are not generally interested in seeing `batch_interrupt_received=False` on their screens and this fact should be somehow reflected in the name of this record.

**Comment by bartvm** [comment]
> There are before_epoch and after_epoch for that.

Those are callbacks, I was talking about e.g. a script that does plotting after training has completed.

> I think that regardless of the frontend used to display training progress people are not generally interested in seeing batch_interrupt_received=False on their screens and this fact should be somehow reflected in the name of this record.

I think you're making my point: You want to apply name mangling in order to serve rendering purposes. I don't think we should chose our naming based on that. W would need to start adding `.trim('_')` to each extension that does want to print these entries for some reason, plus we're introducing all sorts of global conventions. Is it really that bad to just have: `hide=['batch_interrupt_received', 'epoch_interrupt_received'`, ...]`? It's more flexible too. We can let the user override it if he wants, or add to the list in case there are other records that need to be hidden.

**Comment by rizar** [comment]
> Is it really that bad to just have: hide=['batch_interrupt_received', 'epoch_interrupt_received', ...]`?

Maybe I am wrong, but it seems very bad to me. 

Now we have only `Printing`, but assume that we have also `NCursesPrinting` and `BrowserPrinting`. Each of them will have to possess its own default black list. They could share this blacklist, but this would be more like a workaround to hide the general design flow: lack of logging levels.

Furthermore, a single main loop can use up to a few dozens of extensions. Some of them will store their state in `log.status`, since it is the most resumption-friendly way of storing the state. To silent this channels the user will have to collect this-long-black-list-of-them because he can not select them all by a simple condition, such as `s.startswith('_')`.

The white list that you use in your `NotebookPlot` will also grow to tremendous lengths in any real application. Furthermore, the user will now know that a useful record exists because it is not printed because he does not know about it - infinite recursion.

> You want to apply name mangling in order to serve rendering purposes. I don't think we should chose our naming based on that

I want multiple logging levels, and I want them right away as this is an important feature. When we discussed implementing proper logging level you said [here](https://github.com/bartvm/blocks/pull/476/files#r27027221), part of which I cite here `If variables are half-decently named, doing something along the lines of grep cost should be good enough for 99% of the use cases.`. Which I interpret as follows: if we can not afford storing meta-information about log records due to the technical complexity of such an endeavour, it is the record name that has to carry this logging level. In such context a leading underscore is a reasonable solution.

> We would need to start adding .trim('_') to each extension that does want to print these entries for some reason

Why? An underscore is a part of the record name and should be printed. 

> plus we're introducing all sorts of global conventions. 

I agree that the ideal way to go is meta-information for log records stored in the log. But let's leave it for Blocks 0.2,
